packages/flutter/lib/src/material/stepper.dart

```diff
@@ -279,8 +328,65 @@ class Stepper extends StatefulWidget {
   /// }
   /// ```
   /// {@end-tool}
+  @Deprecated(
+    'Migrate to indexedControlsBuilder. '
+    'This builder was deprecated after v2.4.0-1.0.pre.126.',
+  )
   final ControlsWidgetBuilder? controlsBuilder;
 
+  /// The callback for creating custom controls.
+  ///
+  /// If null, the default controls from the current theme will be used.
+  ///
+  /// This callback which takes in a context and a [ControlsDetails] object, which
+  /// contains step information and two functions: [onStepContinue] and [onStepCancel].
+  /// These can be used to control the stepper. For example, reading the
+  /// [ControlsDetails.currentStep] value within the callback can change the text
+  /// of the continue or cancel button depending on which step users are at.
+  ///
+  /// {@tool dartpad --template=stateless_widget_scaffold}
+  /// Creates a stepper control with custom buttons.
+  ///
+  /// ```dart
+  /// Widget build(BuildContext context) {
+  ///   return Stepper(
+  ///     indexedControlsBuilder:
+  ///       (BuildContext context, { ControlsDetails details }) {
+  ///          return Row(
+  ///            children: <Widget>[
+  ///              TextButton(
+  ///                onPressed: details.onStepContinue
+  ///                child: const Text('Continue to Step ${details.stepIndex + 1}'),
+  ///              ),
+  ///              TextButton(
+  ///                onPressed: details.onStepCancel,
+  ///                child: const Text('Back to Step {details.stepIndex - 1}'),
+  ///              ),
+  ///            ],
+  ///          );
+  ///       },
+  ///     steps: const <Step>[
+  ///       Step(
+  ///         title: Text('A'),
+  ///         content: SizedBox(
+  ///           width: 100.0,
+  ///           height: 100.0,
+  ///         ),
+  ///       ),
+  ///       Step(
+  ///         title: Text('B'),
+  ///         content: SizedBox(
+  ///           width: 100.0,
+  ///           height: 100.0,
+  ///         ),
+  ///       ),
+  ///     ],
+  ///   );
+  /// }
+  /// ```
+  /// {@end-tool}
+  final IndexedControlsBuilder? indexedControlsBuilder;
```

Comment:
I do not like this name, but so far have struggled to come up with something very good. `controlsBuilder` was aptly named 😆